<font style='font-size:1.5em'>**💻 Week 07 lab – Simple data cleaning with `pd.apply()`**</font>

<font style='font-size:1.2em'>DS105W – Data for Data Science</font>

**AUTHORS:**  Dr. [Jon Cardoso-Silva](https://jonjoncardoso.github.io) 

**DEPARTMENT:** [LSE Data Science Institute](https://lse.ac.uk/dsi)

**LAST REVISION:** 23 February 2024


--- 

**⚙️ The setup**

We will use the same libraries as we have been using in the previous weeks. 


Run the cell bellow to ensure you have the most up-to-date version of pandas

In [2]:
!pip install --upgrade pandas

In [18]:
import io                     # We need to deal with I/O
import requests               # This is how we access the web
import numpy as np            # This is how we work with arrays
import pandas as pd           # This is how we work with data frames

from pprint import pprint     # Print things in a pretty way
from scrapy import Selector   # This is how we parse HTML
from io import StringIO

**📚 LEARNING OBJECTIVES:***

- Discover the best way to search for Python help online
- Finally understand why custom functions are helpful in pandas
- Learn how to use `pd.apply()` to apply a function to each row of a DataFrame

---

# Part 0: Export your chat logs (~ 3 min)

As part of the <span style="font-weight:bold"> ![](/figures/icons/GENIAL_favicon.png){width=1em} GEN<font color='#D55816'>IA</font>L</span> project, we ask that you fill out the following form as soon as you come to the lab:

🎯 **ACTION POINTS**

1. 🔗 [**CLICK HERE**](https://forms.office.com/e/689MersZzV) to export your chat log.

    Thanks for being GENIAL! 🎟️

::: {style="width:70%;border: 1px solid #aaa; border-radius:1em; padding: 1em; margin-left:1.5em;"}

👉 **NOTE:** You MUST complete the [**initial form**](https://forms.office.com/e/UsYK256Byf).

If you _really_ don't want to participate in GENIAL, just answer 'No' to the Terms & Conditions question - your e-mail address will be deleted from GENIAL's database the following week.

:::

## Join us this Tuesday for the GENIAL Open Lecture:

<iframe src="https://www.lse.ac.uk/DSI/Events/2023-24/GENIAL-Open-Lecture" style="border:0px #ffffff none;" name="myiFrame" scrolling="no" frameborder="1" marginheight="0px" marginwidth="0px" height="350px" width="700px" allowfullscreen></iframe>

---

# Part I: A pandas trick to scrape tables from the web (20 min)


More than learning about a new function, this section is about thinking about a problem and discussing the different ways of researching and making sense of solutions.

**As soon as you get to the class, you can jump straight into the tasks.**

## Task 1: Reading an HTML table into a DataFrame

🎯 **ACTION POINTS:**


1. Access the Wikipedia entry for the [List of national capitals by population](https://en.wikipedia.org/wiki/List_of_national_capitals_by_population) and locate the main table on the page.

2. We will learn how to use the `pd.read_html()` function from the `pandas` library to parse a table directly from HTML into a DataFrame. The code below shows an example of how to do that. All you have to do is replace the `<blank>` placeholder with the appropriate CSS selector for the table we identified in Step 1:

We see that the correct selector is
```html
<table class="wikitable...
```

In python, variables are either objects, their methods or attributes

Method:
e.g.
```python
object.method()
Selector(text)
requests.get(base_url)
```

Attribute:
e.g.
```python
object.attribute
response.text
```


In [27]:
base_url = "https://en.wikipedia.org/wiki/List_of_national_capitals_by_population"

response = requests.get(base_url)
sel = Selector(text=response.text)

# Identify the CSS selector for the '19-th century' table
table_selector = 'table.wikitable'
#better that <tbody>

# Extract the HTML elements matching the CSS selector
table = sel.css(table_selector).extract_first()

#pd.read_html(table) returns a single list of multiple dataframes
#type(pd.read_html(table))
df = pd.read_html(table)[0]
# in the above the read_html is the method and [0] is the index
type(df)

pandas.core.frame.DataFrame

In [21]:
base_url = "https://en.wikipedia.org/wiki/List_of_national_capitals_by_population"


# the below doesn't use our good selector
for table in pd.read_html(base_url):
    display(table)


Lists of capitals
                                        Of countries
0  In alphabetical order By latitude By populatio...
1                            Of country subdivisions
2  Capitals outside the territories they serve Pu...
3                                      Cities portal
4                                                vte

Country / dependency            Capital  \
0                                              China *            Beijing   
1                                              Japan *              Tokyo   
2                                             Russia *             Moscow   
3                                           DR Congo *           Kinshasa   
4                                          Indonesia *            Jakarta   
..                                                 ...                ...   
234                         Norfolk Island (Australia)           Kingston   
235                Cocos (Keeling) Islands (Australia)        West Island   
236                              Pitcairn Islands (UK)          Adamstown   
237  South Georgia and the South Sandwich Islands (UK)  King Edward Point   
238                                              Palau          Ngerulmud   

    Population % of country    Source  
0     21542000         1.5%  [1] 2018  
1     14094034        11.3%  [2] 2023  
2     13104177         9.0%  [3] 2023  
3     12691000        13.2%  [4] 2017  
4     10562088         3.9%  [5] 2020  
..         ...          ...       ...  
234        341          NaN      2015  
235        134        24.6%      2011  
236         40       100.0%      2021  
237         22        73.3%      2018  
238          0         0.0%      2010  

[239 rows x 5 columns]

vteWorld's largest cities  \
0                   City proper   
1             Metropolitan area   
2      Urban area/agglomeration   
3                    Historical   
4              Related articles   
5  Cities portal • World portal   

                         vteWorld's largest cities.1  
0  Capitals Africa Americas North Latin Central S...  
1  Americas (North South West Indies) Europe (Eur...  
2  Asia Africa Europe European Union Nordic North...  
3                                       World Europe  
4  Arcology Ecumenopolis Global city Megacity Meg...  
5                       Cities portal • World portal

vteLists of countries by population statistics  \
0                                            Global   
1                             Continents/subregions   
2                                  Intercontinental   
3                                Cities/urban areas   
4                                   Past and future   
5                                Population density   
6                                 Growth indicators   
7                                   Life expectancy   
8                                Other demographics   
9                                            Health   
10                         Education and innovation   
11                                         Economic   
12  List of international rankings Lists by country   

     vteLists of countries by population statistics.1  
0   Current population United Nations Demographics...  
1   Africa Antarctica Asia Europe North America Ca...  
2   Americas Arab world Commonwealth of Nations Eu...  
3   World cities National capitals Megacities Mega...  
4   Past and future population World population es...  
5   Current density Past and future population den...  
6   Population growth rate Natural increase Net re...  
7   world Africa Asia Europe North America Oceania...  
8   Age at childbearing Age at first marriage Age ...  
9   Antidepressant consumption Antiviral medicatio...  
10  Bloomberg Innovation Index Education Index Int...  
11  Access to financial services Development aid d...  
12    List of international rankings Lists by country

You should see something like:

<div style="width:40%;font-size:0.75em">

```output
[                                  Country / dependency            Capital  \
 0                                              China *            Beijing   
 1                                              Japan *              Tokyo   
 2                                             Russia *             Moscow   
 3                                           DR Congo *           Kinshasa   
 4                                          Indonesia *            Jakarta   
 ..                                                 ...                ...   
 234                         Norfolk Island (Australia)           Kingston   
 235                Cocos (Keeling) Islands (Australia)        West Island   
 236                              Pitcairn Islands (UK)          Adamstown   
 237  South Georgia and the South Sandwich Islands (UK)  King Edward Point   
 238                                              Palau          Ngerulmud   
 
     Population % of country    Source  
 0     21542000         1.5%  [1] 2018  
 1     14094034        11.3%  [2] 2023  
 2     13104177         9.0%  [3] 2023  
 3     12691000        13.2%  [4] 2017  
 4     10562088         3.9%  [5] 2020  
 ..         ...          ...       ...  
 234        341          NaN      2015  
 235        134        24.6%      2011  
 236         40       100.0%      2021  
 237         22        73.3%      2018  
 238          0         0.0%      2010  
 
 [239 rows x 5 columns]]
```

</div>

2. How do you extract the data frame from the output above?

In [22]:
df

Country / dependency            Capital  \
0                                              China *            Beijing   
1                                              Japan *              Tokyo   
2                                             Russia *             Moscow   
3                                           DR Congo *           Kinshasa   
4                                          Indonesia *            Jakarta   
..                                                 ...                ...   
234                         Norfolk Island (Australia)           Kingston   
235                Cocos (Keeling) Islands (Australia)        West Island   
236                              Pitcairn Islands (UK)          Adamstown   
237  South Georgia and the South Sandwich Islands (UK)  King Edward Point   
238                                              Palau          Ngerulmud   

    Population % of country    Source  
0     21542000         1.5%  [1] 2018  
1     14094034        11.3%  [2] 2023  
2     13104177         9.0%  [3] 2023  
3     12691000        13.2%  [4] 2017  
4     10562088         3.9%  [5] 2020  
..         ...          ...       ...  
234        341          NaN      2015  
235        134        24.6%      2011  
236         40       100.0%      2021  
237         22        73.3%      2018  
238          0         0.0%      2010  

[239 rows x 5 columns]

## Task 2: Learning to read warnings + official documentation


If you have pandas 2.1.0 or above installed, you probably saw the following warning when you ran the code above:

<div style="width:50%;font-size:0.9em;">
 
```output
FutureWarning: Passing literal html to 'read_html' is deprecated and will be removed in a future version. To read from a literal string, wrap it in a 'StringIO' object.
  pd.read_html(table)
```

</div>


🎯 **ACTION POINTS:**



1. What is [StringIO](https://docs.python.org/3/library/io.html#io.StringIO) ? Answer in your own words using Markdown bullet points.


It is a method from the io library that enables us to overcome the above error. It provides a text stream with an in-text buffer.

\# Delete this line and write your answer here



2. Rewrite the code above such that it no longer throws a warning. (You might need to do a bit of research on your own)

#One way of doing it
```python
import warnings

warnings.simplefilter(action = 'ignore', category = FutureWarning)
df = pd.read_html(table)[0]
```

In [26]:
table

vteLists of countries by population statistics  \
0                                            Global   
1                             Continents/subregions   
2                                  Intercontinental   
3                                Cities/urban areas   
4                                   Past and future   
5                                Population density   
6                                 Growth indicators   
7                                   Life expectancy   
8                                Other demographics   
9                                            Health   
10                         Education and innovation   
11                                         Economic   
12  List of international rankings Lists by country   

     vteLists of countries by population statistics.1  
0   Current population United Nations Demographics...  
1   Africa Antarctica Asia Europe North America Ca...  
2   Americas Arab world Commonwealth of Nations Eu...  
3   World cities National capitals Megacities Mega...  
4   Past and future population World population es...  
5   Current density Past and future population den...  
6   Population growth rate Natural increase Net re...  
7   world Africa Asia Europe North America Oceania...  
8   Age at childbearing Age at first marriage Age ...  
9   Antidepressant consumption Antiviral medicatio...  
10  Bloomberg Innovation Index Education Index Int...  
11  Access to financial services Development aid d...  
12    List of international rankings Lists by country

In [28]:
df = pd.read_html(StringIO(table))[0]

\# Delete this line and write your answer here

4. 💭 **REFLECTION TIME:** Summarise your troubleshooting process below. Use Markdown bullet points to write your answer.

   - **Finding the source of the problem:** how did you discover what was "wrong" with the code? For example, did you Google the warning message? Did you prompt a chatbot? Did you read the official documentation?

   - **How did you assess the accuracy of your answer?** Or, in other words, how did you know that your solution was correct?

   - **Why does your solution work?** In other words, explain why `pd.read_html` no longer like strings.

\# Delete this line and write your answers here

5. Let's rename columns to mirror the type of names that are most commonly used in Python:
    
    - `Country / dependency` -> `country_dependency` (lower case, underscore instead of space)

    - `Capital` -> `capital`

    - `Population` -> `population`

    - `% of country` -> `country_percentage`
    
    - `Source` -> `source`

In [29]:
# If you already know precisely the order of the columns, you can rename them directly:
df.columns = ['country_dependency', 'capital', 'population', 'country_percentage', 'source']
#th above gives us access to the columns attribute of the dataframe object
#UpperCamelCase for class names if using Spyder
df.columns

Index(['country_dependency', 'capital', 'population', 'country_percentage',
       'source'],
      dtype='object')

::: {style="font-size:0.9em;width:80%;margin-left:2em;"}

The other way, which is more robust in the long term, is using the rename method. Read about it in the documentation [here](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.rename.html).

You would have to write something like:

```python
df.rename(columns={'Country / dependency': 'country_dependency', 
                   'Capital': 'capital', 
                   'Population': 'population', 
                   '% of country': 'country_percentage', 
                   'Source': 'source'}, 
          inplace=True) # If you don't use inplace=True, you will have to assign the result to df
```

:::

## 👨🏻‍🏫 TEACHING MOMENT

Your teacher will mediate a classroom discussion of your responses to the tasks above. Then, they will show a solution to the problem and explain its reasoning.


---

# Part II: Is data cleaning a political act? (40 min)

In the past, we showed you that you can convert an entire column of a DataFrame to a different data type. For example, the `population` column of our data frame is currently a string, but to calculate the average population of capitals, we would have to convert it to an integer first. 

However, if you try to convert the `population` column to an integer using the `astype()` method:

```python
df['population'].astype(np.int32)
```

We get a `ValueError` because the column contains non-numeric characters.

In [30]:
df.population.astype(np.int32)

ValueError: invalid literal for int() with base 10: '217,732 3,000'

In [31]:
#no problem with parsing '_', but decimals are ignored with the int method
print(float(123_123.123))
int(123_123.123)

123123.123


123123

Different methods of identifying the problem:
```python
str.is_numeric()
str.is_alnum()
str.is_alpha()
str.isdigit()
try:
    ____
except:
    _____
```


👥 **WORK ON THESE IN PAIRS:**

Let's start by practising some troubleshooting skills:

1. **Identify the problem**: The error message tells you that the conversion from string to integer failed for a particular population value. What is the value that caused the error?

The two capital cities in the SADR and in one other country. We got:

ValueError: invalid literal for int() with base 10: '217,732 3,000'


2. **Was that the only one?** How would you write code to detect **all** the non-numeric values in the `Population` column? Don't worry about fixing the data frame yet, just write the code to detect the non-numeric values.

<details style="font-size:0.9em;width:40%;margin-left:2em;"><summary>Click here to read a few Python tips</summary>

You can loop through the values of a column in a for loop like this:

```python
for value in df['population']:
    ...do something with value here...
```


</details>

In [32]:
for value in df['population']:
    if value.isnumeric() == False:
        print(value)

217,732 3,000
449 0


In [33]:
for value in df.population:
    try:
        float(value)
    except ValueError:
        print(value)

217,732 3,000
449 0


```python
str(df.population).str.is_numeric()
```
returns an error

3. **Wrap your code in a function:** Use the reasoning behind the code you wrote in Step 2 and create a function called `is_invalid_number` that takes a string as input and returns `True` if the string is not a valid number and `False` otherwise.


    <div style="width:60%;font-size:0.9em;">
    
    💡 **Why bother?** As explained in W05 lecture, we want you to become proficient with custom functions. 
    
    [They are great for doing data manipulation in pandas! More on that in Step 5 of this current action point.]{style="margin-left:1.5em;"}
    
    [THANKS FOR SEEING THIS HIDDEN MARKDOWN PRO-TIP! Did you notice this weird syntax I'm using here? I have text wrapped within square brackets followed by some curly braces... This is a nice little hacky way to apply some inline CSS to the text when writing in Markdown :D ]{style="margin-left:1.5em;display:block;color:#ffffff"}

    </div>

In [34]:
def is_invalid_number(value):
    return not value.isnumeric()

In [ ]:
#def is_invalid_number(value):
    #return not value.isdigit()

You can also use the 
```python 
any(bool)
```
function which returns 
```python 
True
```
if any of the bools are 
```python 
True
```

4. **Test your function:** Test your function with the following inputs:
 
    - `'123'`

    - `'abc'`

    - `'123 784'`
    
    What do you expect the output to be for each of these inputs?

In [35]:
# Delete this line and write your answer here
print(is_invalid_number('123_123123'))
print(is_invalid_number('123'))
print(is_invalid_number('abc'))

True
False
True


::: {style="font-size:0.9em;width:40%;margin-left:2em;"}

**Note**: in real life, we wouldn't need to write a function just for this. There are built-in functions in Python that can do this for us. But we are doing this to practice writing custom functions.

**Pro-tip**: See people fighting online for the 'proper' way to identify whether a string constitutes a number by visiting [this StackOverflow](https://stackoverflow.com/questions/354038/how-do-i-check-if-a-string-represents-a-number-float-or-int) discussion thread.

:::

5. **⭐️ Applying the function to the DataFrame:** This action point will hopefully clarify why we are doing all of this function stuff. 

    Pandas has a method called `apply()` that applies a function to each element of a DataFrame and returns a new column with the results.

    If your function worked well, then the code below should work and return a new column full with `True` and `False` values.

In [36]:
df['population'].apply(is_invalid_number)


0      False
1      False
2      False
3      False
4      False
       ...  
234    False
235    False
236    False
237    False
238    False
Name: population, Length: 239, dtype: bool

6. **Save this to a new column:** Save the results of the `apply()` method to a new column called `is_invalid_pop_number` in the DataFrame.

    This can be achieved by running the following code:

In [37]:
df['is_invalid_pop_number'] = df['population'].apply(is_invalid_number)

7. **Filter the DataFrame:** Use the `is_invalid_pop_number` column to filter the DataFrame to only include rows where `is_invalid_pop_number` is `True`.

In [38]:
# I only want the lines where the population is invalid
selected_rows = df['is_invalid_pop_number'] == True

# I pass the list of booleans to the dataframe to select the rows I want
df[selected_rows]

# pick one of the invalid values and fix it

country_dependency                                 capital  \
152  Sahrawi Arab Democratic Republic  Laayoune (claimed) Tifariti (de facto)   
233                   Montserrat (UK)    Brades (de facto) Plymouth (de jure)   

        population country_percentage source  is_invalid_pop_number  
152  217,732 3,000                  —   2014                   True  
233          449 0                NaN   2011                   True

8. **🏆 How will you solve that?** 

    Say we don't want to remove the two lines above, but instead, we like to keep only one of the two numbers that appear in their `population` column. How would you do that?

    Write the code below that meet the following requirements:

    - Create a custom function that takes a string as input. If the string is a valid number, return the number. Otherwise, choose one of the numbers and return it.
    
    - The function must return a number, not a string passing as a number!

    - Use the `apply()` method to apply the function to the `population` column and save the results to a new column called `clean_population`.

    - Ensure that the `clean_population` is of type `np.int32`

In [39]:
def clean_pop(value):
    if not is_invalid_number(value):
        return int(value)
    else:
        return int(value.split(' ')[0].replace(',', ''))


In [40]:
int(df[selected_rows].iloc[0]['population'].split(' ')[1].replace(',', ''))

3000

In [41]:
df['clean_population'] = df.population.apply(clean_pop)
df['clean_population']

0      21542000
1      14094034
2      13104177
3      12691000
4      10562088
         ...   
234         341
235         134
236          40
237          22
238           0
Name: clean_population, Length: 239, dtype: int64

9. **Calculate the average population:** Now that you have cleaned the `population` column, you can calculate the average population of the capitals by running the code below. 

In [42]:
df['clean_population'].mean()

1623159.3849372384

Or even a more complete statistical summary of the `clean_population` column:

In [43]:
df['clean_population'].describe()

count    2.390000e+02
mean     1.623159e+06
std      2.851528e+06
min      0.000000e+00
25%      7.117400e+04
50%      6.040130e+05
75%      1.749742e+06
max      2.154200e+07
Name: clean_population, dtype: float64

In [44]:
# Or go full fancy and use the pandas style feature to format the output

(
    df['clean_population']
    .describe()         # get the summary statistics
    .to_frame()         # convert the output of the describe method from a series to a data frame
    .transpose()        # transpose the data frame
    .style
    .format("{:,.0f}") # apply the comma format to the numbers
)

---

# Part III – Practice, practice, practice (20 min)

🎯 **ACTION POINTS:**

1. Pre-process the `country_percentage` column to remove the `%` sign and convert it to a float.

2. Then, calculate the average percentage of the country that each capital represents.

In [46]:
def fix_country_percentage(percentage):
    value_without_pct_sign = str(percentage).replace('&', '')
    if value_without_pct_sign == '_':
        return np.nan
    else:
        return float(value_without_pct_sign)
    

In [48]:
def fix_percentage(value):
    number = str(value).replace('&', '')[0]

    try:
        number = float(number)
        return number
    except ValueError:
        return np.nan

In [49]:
df['clean_country_percentage'] = df['country_percentage'].apply(fix_percentage)

In [50]:
df


country_dependency            capital  \
0                                              China *            Beijing   
1                                              Japan *              Tokyo   
2                                             Russia *             Moscow   
3                                           DR Congo *           Kinshasa   
4                                          Indonesia *            Jakarta   
..                                                 ...                ...   
234                         Norfolk Island (Australia)           Kingston   
235                Cocos (Keeling) Islands (Australia)        West Island   
236                              Pitcairn Islands (UK)          Adamstown   
237  South Georgia and the South Sandwich Islands (UK)  King Edward Point   
238                                              Palau          Ngerulmud   

    population country_percentage    source  is_invalid_pop_number  \
0     21542000               1.5%  [1] 2018                  False   
1     14094034              11.3%  [2] 2023                  False   
2     13104177               9.0%  [3] 2023                  False   
3     12691000              13.2%  [4] 2017                  False   
4     10562088               3.9%  [5] 2020                  False   
..         ...                ...       ...                    ...   
234        341                NaN      2015                  False   
235        134              24.6%      2011                  False   
236         40             100.0%      2021                  False   
237         22              73.3%      2018                  False   
238          0               0.0%      2010                  False   

     clean_population  clean_country_percentage  
0            21542000                       1.0  
1            14094034                       1.0  
2            13104177                       9.0  
3            12691000                       1.0  
4            10562088                       3.0  
..                ...                       ...  
234               341                       NaN  
235               134                       2.0  
236                40                       1.0  
237                22                       7.0  
238                 0                       0.0  

[239 rows x 8 columns]